src/AutoRest.CSharp/Generation/Writers/ClientOptionsWriter.cs

```diff
@@ -0,0 +1,91 @@
+
+        public static void WriteClientOptions(CodeWriter writer, BuildContext context)
+        {
+            var clientOptionsName = GetClientOptionsPrefix(context.DefaultLibraryName);
+            var @namespace = context.DefaultNamespace;
+            var apiVersions = context.CodeModel.OperationGroups
+                .SelectMany(g => g.Operations.SelectMany(o => o.ApiVersions))
+                .Select(v => v.Version)
+                .Distinct()
+                .OrderBy(v => v)
+                .Select(v => (Version: v, Name: ToVersionProperty(v)))
+                .ToArray();
+
+            using (writer.Namespace(@namespace))
+            {
+                writer.WriteXmlDocumentationSummary($"Client options for {clientOptionsName}Client.");
+                using (writer.Scope($"public class {clientOptionsName}ClientOptions: {typeof(ClientOptions)}"))
+                {
+                    writer.Line($"private const ServiceVersion LatestVersion = ServiceVersion.{apiVersions.Last().Name};");
+                    writer.Line();
+                    writer.WriteXmlDocumentationSummary("The version of the service to use.");
+                    using (writer.Scope($"public enum ServiceVersion"))
+                    {
+                        int i = 1;
+                        foreach (var apiVersion in apiVersions)
+                        {
+                            writer.WriteXmlDocumentationSummary($"Service version \"{apiVersion.Version}\"");
+                            writer.Line($"{apiVersion.Name} = {i:L},");
+                            i++;
+                        }
+                    }
+
+                    writer.Line();
+                    writer.Line($"internal string Version {{ get; }}");
+                    writer.Line();
+
+                    writer.WriteXmlDocumentationSummary($"Initializes new instance of {clientOptionsName}ClientOptions.");
+                    using (writer.Scope($"public {clientOptionsName}ClientOptions(ServiceVersion version = LatestVersion)"))
+                    {
+                        writer.Append($"Version = version ");
+                        using (writer.Scope($"switch", end: "};"))
+                        {
+                            foreach (var apiVersion in apiVersions)
+                            {
+                                writer.Line($"ServiceVersion.{apiVersion.Name} => {apiVersion.Version:L},");
+                            }
+
+                            writer.Line($"_ => throw new {typeof(NotSupportedException)}()");
+                        }
+                    }
+                }
+            }
+        }
+
+        private static string ToVersionProperty(string s)
+        {
+            return "V" + s.Replace(".", "_").Replace('-', '_');
+        }
+
+        public static string GetClientOptionsPrefix(string name)
+        {
+            if (name.EndsWith(OperationsSuffixValue) && name.Length >= OperationsSuffixValue.Length)
```

Comment:
We seem to start having a lot of copies of this code.

Reply:
Where should we put this common method?

Reply:
Removed this copy and now using the same method from ClientBase.